pl_bolts/models/detection/retinanet/retainanet_module.py

```diff
@@ -0,0 +1,150 @@
+from argparse import ArgumentParser
+from typing import Any, Optional
+
+import pytorch_lightning as pl
+import torch
+
+from pl_bolts.metrics.object_detection import _evaluate_iou
+from pl_bolts.models.detection.retinanet import create_retinanet_backbone
+from pl_bolts.utils import _TORCHVISION_AVAILABLE
+from pl_bolts.utils.warnings import warn_missing_pkg
+
+if _TORCHVISION_AVAILABLE:
+    from torchvision.models.detection.retinanet import RetinaNet as torchvision_RetinaNet
+    from torchvision.models.detection.retinanet import retinanet_resnet50_fpn, RetinaNetHead
+else:  # pragma: no cover
+    warn_missing_pkg("torchvision")
+
+
+class RetinaNet(pl.LightningModule):
+    """
+    PyTorch Lightning implementation of Retina Net `Focal Loss for
+    Dense Object Detection <https://arxiv.org/abs/1708.02002>`_.
+
+    Paper authors: Tsung-Yi Lin, Priya Goyal, Ross Girshick, Kaiming He, Piotr Dollár
+
+    Model implemented by:
+        - `Aditya Oke <https://github.com/oke-aditya>`
+
+    During training, the model expects both the input tensors, as well as targets (list of dictionary), containing:
+        - boxes (`FloatTensor[N, 4]`): the ground truth boxes in `[x1, y1, x2, y2]` format.
+        - labels (`Int64Tensor[N]`): the class label for each ground truh box
+
+    CLI command::
+
+        # PascalVOC
+        python retinanet_module.py --gpus 1 --pretrained True
+    """
+
+    def __init__(
+        self,
+        learning_rate: float = 0.0001,
+        num_classes: int = 91,
+        backbone: Optional[str] = None,
+        fpn: bool = True,
+        pretrained: bool = False,
+        pretrained_backbone: bool = True,
+        trainable_backbone_layers: int = 3,
+        **kwargs: Any,
+    ):
+        """
+        Args:
+            learning_rate: the learning rate
+            num_classes: number of detection classes (including background)
+            backbone: Pretained backbone CNN architecture.
+            fpn: If True, creates a Feature Pyramind Network on top of Resnet based CNNs.
+            pretrained: if true, returns a model pre-trained on COCO train2017
+            pretrained_backbone: if true, returns a model with backbone pre-trained on Imagenet
+            trainable_backbone_layers: number of trainable resnet layers starting from final block
+        """
+        super().__init__()
+        self.learning_rate = learning_rate
+        self.num_classes = num_classes
+        self.backbone = backbone
+        if backbone is None:
```

Comment:
This complexity of different combinations of arguments in the Faster R-CNN and RetinaNet models is quite confusing to me. If `backbone` ("*pretrained backbone CNN architecture*") is not given, one of the predefined Torchvision model architectures is used (`fasterrcnn_resnet50_fpn` or `retinanet_resnet50_fpn`), pretrained model and backbone may still be used, and `trainable_backbone_layers` is ignored. If `backbone` is given, then the `pretrained` argument is ignored.

We could improve the documentation to try and make this clear, but I was thinking that would it be more intuitive to
* make `backbone` a required argument in these models
* remove `pretrained` argument from these models
* add a third model `TorchvisionDetectionModel` that takes as an argument one of the predefined torchvision model names, `fasterrcnn_resnet50_fpn`, `fasterrcnn_mobilenet_v3_large_320_fpn`, `fasterrcnn_mobilenet_v3_large_fpn`, `retinanet_resnet50_fpn`, `keypointrcnn_resnet50_fpn`, `maskrcnn_resnet50_fpn`, etc. and supports also loading an entirely pretrained model.

Reply:
Backbones have a problem. Not all backbones work with FPN.
Not all have pretrained weights on COCO available, some have on imagenet. 
Pretrained argument is needed, so we can't remove it.
Pretrainde either load COCO weights or ImageNet weights.

This can be made explicit making pretrained an 'str' which can be = 'imagenet' or 'coco')
But this was deferred for #200 .

Current API works as follows.
The reason for `pretraind_backbone_layers` being ingored is [here](https://github.com/pytorch/vision/issues/3406)
Since it is fixed, we can pass that param. But again this needs a mandatory torchvision bump which cannot be enforced.
See #554 .

Current API works as

```
model = RetinaNet(backbone="resnet50", fpn=True)
```

We can instead refactor as follows: -
All detection models accept backbone as `nn.Module` not an `str`.
Users should create a backbone using the method provided to `create_xyz_backbone`.

E.g.  We have provided `create_retinanet_backbone`

```


backbone_model = retinanet.create_retinanet_backbone(backbone, fpn, pretrained_backbone, **kwargs)

Here backbone is "str", so user can pass "resnet50", "mobilenet" etc.
We can easily load all the above supported models.

model = RetinaNet(backbone_model, num_classes, learning_rate, **kwargs)

```

This separates both backbone creation and model creation, plus offers a clear API for all models.

To directly access pre trained models, users can simply create them.

`TorchvisionDetectionModel` is a fine abstraction, but do we need to complicate it ?
User would like to subclass RetinaNet and continue his customization, hence I feel to avoid this.

Also cc @kaushikb11 since flash too has a similar scenario to handle backbones.

Reply:
`pretrained` argument is needed only when `backbone` is not given. What I meant is that these models would work as if `backbone` is always given (and pretrained models cannot be accessed), whereas `TorchvisionDetectionModel` could load pretrained models.

I like your suggestion of separating backbone and model creation. The user could explicitly create a pretrained backbone. But I don't see how the user could access pretrained model weights.

Reply:
I'm not sure if I understood `pretrained` problem.
But this handles as follows (similar to torchvision's handling of `pretrained_backbone` and `pretrained` params)

For backbones, the pretrained weights could be loaded from imagenet if pretrained is passed `true` there.

And if the user passes pretrained in model `RetinaNet` we will try to load `COCO` weights.

Passing both true would not work (it does not in torchvision too you cannot pass `pretrained_backbone` and `pretrained` as True)

E.g. to load ImageNet weights and create pretrained FRCNN / RetinaNet

```
backbone = retinanet.create_retinanet_backbone(backbone="resnet50", fpn, pretrained_backbone=True,,  **kwargs)
model = RetinaNet(backbone, learning_rate, pretrained=False, **kwargs)
```

To load pretrained COCO weights (if available, very few combinations have it)

```
backbone = retinanet.create_backbone(backbone="resnet50", fpn=True, pretrained_backbone=False)
model = RetinaNet(backbone, pretrained=True, num_classes)
```

Note that COCO weights are only available in 3 combinations
`resnet50_fpn`, `mobilnet_large_fpn`, `mobilenet_320_fpn`.

For other cases, the user should initialize with imagenet pretrained backbone (preferable).

Yes, backbone is of course always necessary to create a model. But if we split up backbone creation and model creation it would be easier.

Currently the user has to pass a lot of params to a single Model Definition, which is creating confusion.

Reply:
In your proposal, the backbone argument can be any `nn.Module`, right? I thought it might be tricky then to know which weights to load, if the user creates RetinaNet with `pretrained=True`. But it sounds like you've thought it out. That would indeed be the most intuitive interface!

Reply:
Great. Nice to discuss and comeup with solution! Great ideas @senarvi. 
I will once request @akihironitta  @Borda  to confirm. I will refactor the code accordingly!

Also cc @kaushikb11  if flash has better solutions / ideas or anything if we missed.